File: core/src/main/java/at/stefangeyer/challonge/Challonge.java
package at.stefangeyer.challonge;

import at.stefangeyer.challonge.async.Callback;
import at.stefangeyer.challonge.exception.DataAccessException;
import at.stefangeyer.challonge.model.*;
import at.stefangeyer.challonge.model.enumeration.MatchState;
import at.stefangeyer.challonge.model.enumeration.TournamentType;
import at.stefangeyer.challonge.model.query.AttachmentQuery;
import at.stefangeyer.challonge.model.query.MatchQuery;
import at.stefangeyer.challonge.model.query.ParticipantQuery;
import at.stefangeyer.challonge.model.query.TournamentQuery;
import at.stefangeyer.challonge.model.query.enumeration.TournamentQueryState;
import at.stefangeyer.challonge.rest.RestClient;
import at.stefangeyer.challonge.serializer.Serializer;
import at.stefangeyer.challonge.service.*;
import at.stefangeyer.challonge.service.implementation.SimpleChallongeService;

import java.time.OffsetDateTime;
import java.util.List;

public class Challonge implements ChallongeService {

    private ChallongeService service;

    public Challonge(Credentials credentials, Serializer serializer, RestClient restClient) {
        // Initialize factory
        restClient.initialize(credentials, serializer);

        this.service = new SimpleChallongeService(restClient);
    }

    public final List<Tournament> getTournaments(TournamentQueryState state, TournamentType type, OffsetDateTime createdAfter,
                                                 OffsetDateTime createdBefore, String subdomain) throws DataAccessException {
        return this.service.getTournaments(state, type, createdAfter, createdBefore, subdomain);
    }

    public final List<Tournament> getTournaments() throws DataAccessException {
        return getTournaments(null, null, null, null, null);
    }

    public final void getTournaments(TournamentQueryState state, TournamentType type, OffsetDateTime createdAfter,
                                     OffsetDateTime createdBefore, String subdomain, Callback<List<Tournament>> onSuccess,
                                     Callback<DataAccessException> onFailure) {
        this.service.getTournaments(state, type, createdAfter, createdBefore, subdomain, onSuccess, onFailure);
    }

    public final void getTournaments(TournamentQueryState state, TournamentType type, OffsetDateTime createdAfter,
                                     OffsetDateTime createdBefore, Callback<List<Tournament>> onSuccess,
                                     Callback<DataAccessException> onFailure) {
        getTournaments(state, type, createdAfter, createdBefore, null, onSuccess, onFailure);
    }

    public final void getTournaments(Callback<List<Tournament>> onSuccess, Callback<DataAccessException> onFailure) {
        getTournaments(null, null, null, null, null, onSuccess, onFailure);
    }

    public final Tournament getTournament(String tournament, boolean includeParticipants,
                                          boolean includeMatches) throws DataAccessException {
        return this.service.getTournament(tournament, includeParticipants, includeMatches);
    }

    public final Tournament getTournament(String tournament) throws DataAccessException {
        return getTournament(tournament, false, false);
    }

    public final void getTournament(String tournament, boolean includeParticipants, boolean includeMatches,
                                    Callback<Tournament> onSuccess, Callback<DataAccessException> onFailure) {
        this.service.getTournament(tournament, includeParticipants, includeMatches, onSuccess, onFailure);
    }

    public final void getTournament(String tournament, Callback<Tournament> onSuccess, Callback<DataAccessException> onFailure) {
        getTournament(tournament, false, false, onSuccess, onFailure);
    }

    public final Tournament createTournament(TournamentQuery data) throws DataAccessException {
        return this.service.createTournament(data);
    }

    public final void createTournament(TournamentQuery data, Callback<Tournament> onSuccess,
                                       Callback<DataAccessException> onFailure) {
        this.service.createTournament(data, onSuccess, onFailure);
    }

    public final Tournament updateTournament(Tournament tournament, TournamentQuery data) throws DataAccessException {
        return this.service.updateTournament(tournament, data);
    }

    public final void updateTournament(Tournament tournament, TournamentQuery data, Callback<Tournament> onSuccess,
                                       Callback<DataAccessException> onFailure) {
        this.service.updateTournament(tournament, data, onSuccess, onFailure);
    }

    public final Tournament deleteTournament(Tournament tournament) throws DataAccessException {
        return this.service.deleteTournament(tournament);
    }

    public final void deleteTournament(Tournament tournament, Callback<Tournament> onSuccess,
                                       Callback<DataAccessException> onFailure) {
        this.service.deleteTournament(tournament, onSuccess, onFailure);
    }

    public final Tournament processCheckIns(Tournament tournament, boolean includeParticipants,
                                            boolean includeMatches) throws DataAccessException {
        return this.service.processCheckIns(tournament, includeParticipants, includeMatches);
    }

    public final Tournament processCheckIns(Tournament tournament) throws DataAccessException {
        return processCheckIns(tournament, false, false);
    }

    public final void processCheckIns(Tournament tournament, boolean includeParticipants, boolean includeMatches,
                                      Callback<Tournament> onSuccess, Callback<DataAccessException> onFailure) {
        this.service.processCheckIns(tournament, includeParticipants, includeMatches, onSuccess, onFailure);
    }

    public final void processCheckIns(Tournament tournament, Callback<Tournament> onSuccess,
                                      Callback<DataAccessException> onFailure) {
        processCheckIns(tournament, false, false, onSuccess, onFailure);
    }

    public final Tournament abortCheckIn(Tournament tournament, boolean includeParticipants,
                                         boolean includeMatches) throws DataAccessException {
        return this.service.abortCheckIn(tournament, includeParticipants, includeMatches);
    }

    public final Tournament abortCheckIn(Tournament tournament) throws DataAccessException {
        return abortCheckIn(tournament, false, false);
    }

    public final void abortCheckIn(Tournament tournament, boolean includeParticipants, boolean includeMatches,
                                   Callback<Tournament> onSuccess, Callback<DataAccessException> onFailure) {
        this.service.abortCheckIn(tournament, includeParticipants, includeMatches, onSuccess, onFailure);
    }

    public final void abortCheckIn(Tournament tournament, Callback<Tournament> onSuccess,
                                   Callback<DataAccessException> onFailure) {
        abortCheckIn(tournament, false, false, onSuccess, onFailure);
    }

    public final Tournament startTournament(Tournament tournament, boolean includeParticipants,
                                            boolean includeMatches) throws DataAccessException {
        return this.service.startTournament(tournament, includeParticipants, includeMatches);
    }

    public final Tournament startTournament(Tournament tournament) throws DataAccessException {
        return startTournament(tournament, false, false);
    }

    public final void startTournament(Tournament tournament, boolean includeParticipants, boolean includeMatches,
                                      Callback<Tournament> onSuccess, Callback<DataAccessException> onFailure) {
        this.service.startTournament(tournament, includeParticipants, includeMatches, onSuccess, onFailure);
    }

    public final void startTournament(Tournament tournament, Callback<Tournament> onSuccess,
                                      Callback<DataAccessException> onFailure) {
        startTournament(tournament, false, false, onSuccess, onFailure);
    }

    public final Tournament finalizeTournament(Tournament tournament, boolean includeParticipants,
                                               boolean includeMatches) throws DataAccessException {
        return this.service.finalizeTournament(tournament, includeParticipants, includeMatches);
    }

    public final Tournament finalizeTournament(Tournament tournament) throws DataAccessException {
        return finalizeTournament(tournament, false, false);
    }

    public final void finalizeTournament(Tournament tournament, boolean includeParticipants, boolean includeMatches,
                                         Callback<Tournament> onSuccess, Callback<DataAccessException> onFailure) {
        this.service.finalizeTournament(tournament, includeParticipants, includeMatches, onSuccess, onFailure);
    }

    public final void finalizeTournament(Tournament tournament, Callback<Tournament> onSuccess,
                                         Callback<DataAccessException> onFailure) {
        finalizeTournament(tournament, false, false, onSuccess, onFailure);
    }

    public final Tournament resetTournament(Tournament tournament, boolean includeParticipants,
                                            boolean includeMatches) throws DataAccessException {
        return this.service.resetTournament(tournament, includeParticipants, includeMatches);
    }

    public final Tournament resetTournament(Tournament tournament) throws DataAccessException {
        return resetTournament(tournament, false, false);
    }

    public final void resetTournament(Tournament tournament, boolean includeParticipants, boolean includeMatches,
                                      Callback<Tournament> onSuccess, Callback<DataAccessException> onFailure) {
        this.service.resetTournament(tournament, includeParticipants, includeMatches, onSuccess, onFailure);
    }

    public final void resetTournament(Tournament tournament, Callback<Tournament> onSuccess,
                                      Callback<DataAccessException> onFailure) {
        resetTournament(tournament, false, false, onSuccess, onFailure);
    }

    public final Tournament openTournamentForPredictions(Tournament tournament, boolean includeParticipants,
                                                         boolean includeMatches) throws DataAccessException {
        return this.service.openTournamentForPredictions(tournament, includeParticipants, includeMatches);
    }

    public final Tournament openTournamentForPredictions(Tournament tournament) throws DataAccessException {
        return openTournamentForPredictions(tournament, false, false);
    }

    public final void openTournamentForPredictions(Tournament tournament, boolean includeParticipants,
                                                   boolean includeMatches, Callback<Tournament> onSuccess,
                                                   Callback<DataAccessException> onFailure) {
        this.service.openTournamentForPredictions(tournament, includeParticipants, includeMatches, onSuccess, onFailure);
    }

    public final void openTournamentForPredictions(Tournament tournament, Callback<Tournament> onSuccess,
                                                   Callback<DataAccessException> onFailure) {
        openTournamentForPredictions(tournament, false, false, onSuccess, onFailure);
    }

    public final List<Participant> getParticipants(Tournament tournament) throws DataAccessException {
        return this.service.getParticipants(tournament);
    }

    public final void getParticipants(Tournament tournament, Callback<List<Participant>> onSuccess,
                                      Callback<DataAccessException> onFailure) {
        this.service.getParticipants(tournament, onSuccess, onFailure);
    }

    public final Participant getParticipant(Tournament tournament, long participantId,
                                            boolean includeMatches) throws DataAccessException {
        return this.service.getParticipant(tournament, participantId, includeMatches);
    }

    public final Participant getParticipant(Tournament tournament, long participantId) throws DataAccessException {
        return this.service.getParticipant(tournament, participantId, false);
    }

    public final void getParticipant(Tournament tournament, long participantId, boolean includeMatches,
                                     Callback<Participant> onSuccess, Callback<DataAccessException> onFailure) {
        this.service.getParticipant(tournament, participantId, includeMatches, onSuccess, onFailure);
    }

    public final void getParticipant(Tournament tournament, long participantId, Callback<Participant> onSuccess,
                                     Callback<DataAccessException> onFailure) {
        this.service.getParticipant(tournament, participantId, false, onSuccess, onFailure);
    }

    public final Participant addParticipant(Tournament tournament, ParticipantQuery data) throws DataAccessException {
        return this.service.addParticipant(tournament, data);
    }

    public final void addParticipant(Tournament tournament, ParticipantQuery data, Callback<Participant> onSuccess,
                                     Callback<DataAccessException> onFailure) {
        this.service.addParticipant(tournament, data, onSuccess, onFailure);
    }

    public final List<Participant> bulkAddParticipants(Tournament tournament, List<ParticipantQuery> data) throws DataAccessException {
        return this.service.bulkAddParticipants(tournament, data);
    }

    public final void bulkAddParticipants(Tournament tournament, List<ParticipantQuery> data,
                                          Callback<List<Participant>> onSuccess,
                                          Callback<DataAccessException> onFailure) {
        this.service.bulkAddParticipants(tournament, data, onSuccess, onFailure);
    }

    public final Participant updateParticipant(Participant participant, ParticipantQuery data) throws DataAccessException {
        return this.service.updateParticipant(participant, data);
    }

    public final void updateParticipant(Participant participant, ParticipantQuery data, Callback<Participant> onSuccess,
                                        Callback<DataAccessException> onFailure) {
        this.service.updateParticipant(participant, data, onSuccess, onFailure);
    }

    public final Participant checkInParticipant(Participant participant) throws DataAccessException {
        return this.service.checkInParticipant(participant);
    }

    public final void checkInParticipant(Participant participant, Callback<Participant> onSuccess,
                                         Callback<DataAccessException> onFailure) {
        this.service.checkInParticipant(participant, onSuccess, onFailure);
    }

    public final Participant undoCheckInParticipant(Participant participant) throws DataAccessException {
        return this.service.undoCheckInParticipant(participant);
    }

    public final void undoCheckInParticipant(Participant participant, Callback<Participant> onSuccess,
                                             Callback<DataAccessException> onFailure) {
        this.service.undoCheckInParticipant(participant, onSuccess, onFailure);
    }

    public final Participant deleteParticipant(Participant participant) throws DataAccessException {
        return this.service.deleteParticipant(participant);
    }

    public final void deleteParticipant(Participant participant, Callback<Participant> onSuccess,
                                        Callback<DataAccessException> onFailure) {
        this.service.deleteParticipant(participant, onSuccess, onFailure);
    }

    public final List<Participant> randomizeParticipants(Tournament tournament) throws DataAccessException {
        return this.service.randomizeParticipants(tournament);
    }

    public final void randomizeParticipants(Tournament tournament, Callback<List<Participant>> onSuccess,
                                            Callback<DataAccessException> onFailure) {
        this.service.randomizeParticipants(tournament, onSuccess, onFailure);
    }

    public final List<Match> getMatches(Tournament tournament, Participant participant,
                                        MatchState state) throws DataAccessException {
        return this.service.getMatches(tournament, participant, state);
    }

    public final List<Match> getMatches(Tournament tournament, Participant participant) throws DataAccessException {
        return getMatches(tournament, participant, null);
    }

    public final List<Match> getMatches(Tournament tournament) throws DataAccessException {
        return getMatches(tournament, (Participant) null, null);
    }

    public final void getMatches(Tournament tournament, Participant participant, MatchState state,
                                 Callback<List<Match>> onSuccess, Callback<DataAccessException> onFailure) {
        this.service.getMatches(tournament, participant, state, onSuccess, onFailure);
    }

    public final void getMatches(Tournament tournament, Participant participant, Callback<List<Match>> onSuccess,
                                 Callback<DataAccessException> onFailure) {
        getMatches(tournament, participant, null, onSuccess, onFailure);
    }

    public final void getMatches(Tournament tournament, Callback<List<Match>> onSuccess,
                                 Callback<DataAccessException> onFailure) {
        getMatches(tournament, null, null, onSuccess, onFailure);
    }

    public final Match getMatch(Tournament tournament, long matchId, boolean includeAttachments) throws DataAccessException {
        return this.service.getMatch(tournament, matchId, includeAttachments);
    }

    public final Match getMatch(Tournament tournament, long matchId) throws DataAccessException {
        return getMatch(tournament, matchId, false);
    }

    public final void getMatch(Tournament tournament, long matchId, boolean includeAttachments, Callback<Match> onSuccess,
                               Callback<DataAccessException> onFailure) {
        this.service.getMatch(tournament, matchId, includeAttachments, onSuccess, onFailure);
    }


    public final void getMatch(Tournament tournament, long matchId, Callback<Match> onSuccess,
                               Callback<DataAccessException> onFailure) {
        getMatch(tournament, matchId, false, onSuccess, onFailure);
    }

    public final Match updateMatch(Match match, MatchQuery data) throws DataAccessException {
        return this.service.updateMatch(match, data);
    }

    public final void updateMatch(Match match, MatchQuery data, Callback<Match> onSuccess,
                                  Callback<DataAccessException> onFailure) {
        this.service.updateMatch(match, data, onSuccess, onFailure);
    }

    public final Match markMatchAsUnderway(Match match) throws DataAccessException {
        return this.service.markMatchAsUnderway(match);
    }

    public final void markMatchAsUnderway(Match match, Callback<Match> onSuccess, Callback<DataAccessException> onFailure) {
        this.service.markMatchAsUnderway(match, onSuccess, onFailure);
    }

    public final Match unmarkMatchAsUnderway(Match match) throws DataAccessException {
        return this.service.unmarkMatchAsUnderway(match);
    }

    public final void unmarkMatchAsUnderway(Match match, Callback<Match> onSuccess, Callback<DataAccessException> onFailure) {
        this.service.unmarkMatchAsUnderway(match, onSuccess, onFailure);
    }

    public final Match reopenMatch(Match match) throws DataAccessException {
        return this.service.reopenMatch(match);
    }

    public final void reopenMatch(Match match, Callback<Match> onSuccess, Callback<DataAccessException> onFailure) {
        this.service.reopenMatch(match, onSuccess, onFailure);
    }

    public final List<Attachment> getAttachments(Match match) throws DataAccessException {
        return this.service.getAttachments(match);
    }

    public final void getAttachments(Match match, Callback<List<Attachment>> onSuccess, Callback<DataAccessException> onFailure) {
        this.service.getAttachments(match, onSuccess, onFailure);
    }

    public final Attachment getAttachment(Match match, long attachmentId) throws DataAccessException {
        return this.service.getAttachment(match, attachmentId);
    }

    public final void getAttachment(Match match, long attachmentId, Callback<Attachment> onSuccess,
                                    Callback<DataAccessException> onFailure) {
        this.service.getAttachment(match, attachmentId, onSuccess, onFailure);
    }

    public final Attachment createAttachment(Match match, AttachmentQuery data) throws DataAccessException {
        return this.service.createAttachment(match, data);
    }

    public final void createAttachment(Match match, AttachmentQuery data, Callback<Attachment> onSuccess,
                                       Callback<DataAccessException> onFailure) {
        this.service.createAttachment(match, data, onSuccess, onFailure);
    }

    public final Attachment updateAttachment(Match match, Attachment attachment, AttachmentQuery data) throws DataAccessException {
        return this.service.updateAttachment(match, attachment, data);
    }

    public final void updateAttachment(Match match, Attachment attachment, AttachmentQuery data, Callback<Attachment> onSuccess,
                                       Callback<DataAccessException> onFailure) {
        this.service.updateAttachment(match, attachment, data, onSuccess, onFailure);
    }

    public final Attachment deleteAttachment(Match match, Attachment attachment) throws DataAccessException {
        return this.service.deleteAttachment(match, attachment);
    }

    public final void deleteAttachment(Match match, Attachment attachment, Callback<Attachment> onSuccess,
                                       Callback<DataAccessException> onFailure) {
        this.service.deleteAttachment(match, attachment, onSuccess, onFailure);
    }
}

File: core/src/main/kotlin/at/stefangeyer/challonge/Challonge.kt
package at.stefangeyer.challonge

import at.stefangeyer.challonge.async.Callback
import at.stefangeyer.challonge.exception.DataAccessException
import at.stefangeyer.challonge.model.*
import at.stefangeyer.challonge.model.enum.MatchState
import at.stefangeyer.challonge.model.enum.TournamentType
import at.stefangeyer.challonge.model.query.AttachmentQuery
import at.stefangeyer.challonge.model.query.MatchQuery
import at.stefangeyer.challonge.model.query.ParticipantQuery
import at.stefangeyer.challonge.model.query.TournamentQuery
import at.stefangeyer.challonge.model.query.enum.TournamentQueryState
import at.stefangeyer.challonge.rest.RestClient
import at.stefangeyer.challonge.serializer.Serializer
import at.stefangeyer.challonge.service.AttachmentService
import at.stefangeyer.challonge.service.MatchService
import at.stefangeyer.challonge.service.ParticipantService
import at.stefangeyer.challonge.service.TournamentService
import at.stefangeyer.challonge.service.implementation.SimpleAttachmentService
import at.stefangeyer.challonge.service.implementation.SimpleMatchService
import at.stefangeyer.challonge.service.implementation.SimpleParticipantService
import at.stefangeyer.challonge.service.implementation.SimpleTournamentService
import java.time.OffsetDateTime

class Challonge(credentials: Credentials, serializer: Serializer, private val restClient: RestClient) {

    private val tournaments: TournamentService
    private val participants: ParticipantService
    private val matches: MatchService
    private val attachments: AttachmentService

    init {
        // Initialize factory
        this.restClient.initialize(credentials, serializer)

        this.tournaments = SimpleTournamentService(this.restClient.createTournamentRestClient())
        this.participants = SimpleParticipantService(this.restClient.createParticipantRestClient())
        this.matches = SimpleMatchService(this.restClient.createMatchRestClient())
        this.attachments = SimpleAttachmentService(this.restClient.createAttachmentRestClient())
    }

    /**
     * @see TournamentService
     */
    @JvmOverloads
    @Throws(DataAccessException::class)
    fun getTournaments(state: TournamentQueryState? = null, type: TournamentType? = null, createdAfter: OffsetDateTime? = null,
                       createdBefore: OffsetDateTime? = null, subdomain: String? = null): List<Tournament> =
            this.tournaments.getTournaments(state, type, createdAfter, createdBefore, subdomain)

    /**
     * @see TournamentService
     */
    @JvmOverloads
    fun getTournaments(state: TournamentQueryState? = null, type: TournamentType? = null, createdAfter: OffsetDateTime? = null,
                       createdBefore: OffsetDateTime? = null, subdomain: String? = null, onSuccess: Callback<List<Tournament>>,
                       onFailure: Callback<DataAccessException>) =
            this.tournaments.getTournaments(state, type, createdAfter, createdBefore, subdomain, onSuccess, onFailure)

    /**
     * @see TournamentService
     */
    @JvmOverloads
    @Throws(DataAccessException::class)
    fun getTournament(tournament: String, includeParticipants: Boolean = false, includeMatches: Boolean = false): Tournament =
            this.tournaments.getTournament(tournament, includeParticipants, includeMatches)

    /**
     * @see TournamentService
     */
    @JvmOverloads
    fun getTournament(tournament: String, includeParticipants: Boolean = false, includeMatches: Boolean = false,
                      onSuccess: Callback<Tournament>, onFailure: Callback<DataAccessException>) =
        this.tournaments.getTournament(tournament, includeParticipants, includeMatches, onSuccess, onFailure)

    /**
     * @see TournamentService
     */
    @Throws(DataAccessException::class)
    fun createTournament(data: TournamentQuery): Tournament = this.tournaments.createTournament(data)

    /**
     * @see TournamentService
     */
    fun createTournament(data: TournamentQuery, onSuccess: Callback<Tournament>, onFailure: Callback<DataAccessException>) =
            this.tournaments.createTournament(data, onSuccess, onFailure)

    /**
     * @see TournamentService
     */
    @Throws(DataAccessException::class)
    fun updateTournament(tournament: Tournament, data: TournamentQuery): Tournament =
            this.tournaments.updateTournament(tournament, data)

    /**
     * @see TournamentService
     */
    fun updateTournament(tournament: Tournament, data: TournamentQuery, onSuccess: Callback<Tournament>,
                         onFailure: Callback<DataAccessException>) =
            this.tournaments.updateTournament(tournament, data, onSuccess, onFailure)

    /**
     * @see TournamentService
     */
    @Throws(DataAccessException::class)
    fun deleteTournament(tournament: Tournament): Tournament = this.tournaments.deleteTournament(tournament)

    /**
     * @see TournamentService
     */
    fun deleteTournament(tournament: Tournament, onSuccess: Callback<Tournament>, onFailure: Callback<DataAccessException>) =
            this.tournaments.deleteTournament(tournament, onSuccess, onFailure)

    /**
     * @see TournamentService
     */
    @JvmOverloads
    @Throws(DataAccessException::class)
    fun processCheckIns(tournament: Tournament, includeParticipants: Boolean = false, includeMatches: Boolean = false): Tournament =
            this.tournaments.processCheckIns(tournament, includeParticipants, includeMatches)

    /**
     * @see TournamentService
     */
    @JvmOverloads
    fun processCheckIns(tournament: Tournament, includeParticipants: Boolean = false, includeMatches: Boolean = false,
                        onSuccess: Callback<Tournament>, onFailure: Callback<DataAccessException>) =
            this.tournaments.processCheckIns(tournament, includeParticipants, includeMatches, onSuccess, onFailure)

    /**
     * @see TournamentService
     */
    @JvmOverloads
    @Throws(DataAccessException::class)
    fun abortCheckIn(tournament: Tournament, includeParticipants: Boolean = false, includeMatches: Boolean = false): Tournament =
            this.tournaments.abortCheckIn(tournament, includeParticipants, includeMatches)

    /**
     * @see TournamentService
     */
    @JvmOverloads
    fun abortCheckIn(tournament: Tournament, includeParticipants: Boolean = false, includeMatches: Boolean = false,
                     onSuccess: Callback<Tournament>, onFailure: Callback<DataAccessException>) =
            this.tournaments.abortCheckIn(tournament, includeParticipants, includeMatches, onSuccess, onFailure)

    /**
     * @see TournamentService
     */
    @JvmOverloads
    @Throws(DataAccessException::class)
    fun startTournament(tournament: Tournament, includeParticipants: Boolean = false, includeMatches: Boolean = false): Tournament =
            this.tournaments.startTournament(tournament, includeParticipants, includeMatches)

    /**
     * @see TournamentService
     */
    @JvmOverloads
    fun startTournament(tournament: Tournament, includeParticipants: Boolean = false, includeMatches: Boolean = false,
                        onSuccess: Callback<Tournament>, onFailure: Callback<DataAccessException>) =
            this.tournaments.startTournament(tournament, includeParticipants, includeMatches, onSuccess, onFailure)

    /**
     * @see TournamentService
     */
    @JvmOverloads
    @Throws(DataAccessException::class)
    fun finalizeTournament(tournament: Tournament, includeParticipants: Boolean = false, includeMatches: Boolean = false): Tournament =
            this.tournaments.finalizeTournament(tournament, includeParticipants, includeMatches)

    /**
     * @see TournamentService
     */
    @JvmOverloads
    fun finalizeTournament(tournament: Tournament, includeParticipants: Boolean = false, includeMatches: Boolean = false,
                           onSuccess: Callback<Tournament>, onFailure: Callback<DataAccessException>) =
            this.tournaments.finalizeTournament(tournament, includeParticipants, includeMatches, onSuccess, onFailure)

    /**
     * @see TournamentService
     */
    @JvmOverloads
    @Throws(DataAccessException::class)
    fun resetTournament(tournament: Tournament, includeParticipants: Boolean = false, includeMatches: Boolean = false): Tournament =
            this.tournaments.resetTournament(tournament, includeParticipants, includeMatches)

    /**
     * @see TournamentService
     */
    @JvmOverloads
    fun resetTournament(tournament: Tournament, includeParticipants: Boolean = false, includeMatches: Boolean = false,
                        onSuccess: Callback<Tournament>, onFailure: Callback<DataAccessException>) =
            this.tournaments.resetTournament(tournament, includeParticipants, includeMatches, onSuccess, onFailure)

    /**
     * @see TournamentService
     */
    @JvmOverloads
    @Throws(DataAccessException::class)
    fun openTournamentForPredictions(tournament: Tournament, includeParticipants: Boolean = false, includeMatches: Boolean = false): Tournament =
            this.tournaments.openTournamentForPredictions(tournament, includeParticipants, includeMatches)

    /**
     * @see TournamentService
     */
    @JvmOverloads
    fun openTournamentForPredictions(tournament: Tournament, includeParticipants: Boolean = false, includeMatches: Boolean = false,
                                     onSuccess: Callback<Tournament>, onFailure: Callback<DataAccessException>) =
            this.tournaments.openTournamentForPredictions(tournament, includeParticipants, includeMatches, onSuccess, onFailure)

    /**
     * @see ParticipantService
     */
    @Throws(DataAccessException::class)
    fun getParticipants(tournament: Tournament): List<Participant> =
            this.participants.getParticipants(tournament)

    /**
     * @see ParticipantService
     */
    fun getParticipants(tournament: Tournament, onSuccess: Callback<List<Participant>>, onFailure: Callback<DataAccessException>) =
            this.participants.getParticipants(tournament, onSuccess, onFailure)

    @JvmOverloads
    @Throws(DataAccessException::class)
    fun getParticipant(tournament: Tournament, participantId: Long, includeMatches: Boolean = false): Participant =
            this.participants.getParticipant(tournament, participantId, includeMatches)

    /**
     * @see ParticipantService
     */
    @JvmOverloads
    fun getParticipant(tournament: Tournament, participantId: Long, includeMatches: Boolean = false,
                       onSuccess: Callback<Participant>, onFailure: Callback<DataAccessException>) =
            this.participants.getParticipant(tournament, participantId, includeMatches, onSuccess, onFailure)

    /**
     * @see ParticipantService
     */
    @Throws(DataAccessException::class)
    fun addParticipant(tournament: Tournament, data: ParticipantQuery): Participant =
            this.participants.addParticipant(tournament, data)

    /**
     * @see ParticipantService
     */
    fun addParticipant(tournament: Tournament, data: ParticipantQuery, onSuccess: Callback<Participant>,
                       onFailure: Callback<DataAccessException>) =
            this.participants.addParticipant(tournament, data, onSuccess, onFailure)

    /**
     * @see ParticipantService
     */
    @Throws(DataAccessException::class)
    fun bulkAddParticipants(tournament: Tournament, data: List<ParticipantQuery>): List<Participant> =
            this.participants.bulkAddParticipants(tournament, data)

    /**
     * @see ParticipantService
     */
    fun bulkAddParticipants(tournament: Tournament, data: List<ParticipantQuery>, onSuccess: Callback<List<Participant>>,
                            onFailure: Callback<DataAccessException>) =
            this.participants.bulkAddParticipants(tournament, data, onSuccess, onFailure)

    /**
     * @see ParticipantService
     */
    @Throws(DataAccessException::class)
    fun updateParticipant(participant: Participant, data: ParticipantQuery): Participant =
            this.participants.updateParticipant(participant, data)

    /**
     * @see ParticipantService
     */
    fun updateParticipant(participant: Participant, data: ParticipantQuery, onSuccess: Callback<Participant>,
                          onFailure: Callback<DataAccessException>) =
            this.participants.updateParticipant(participant, data, onSuccess, onFailure)

    /**
     * @see ParticipantService
     */
    @Throws(DataAccessException::class)
    fun checkInParticipant(participant: Participant): Participant =
            this.participants.checkInParticipant(participant)

    /**
     * @see ParticipantService
     */
    fun checkInParticipant(participant: Participant, onSuccess: Callback<Participant>, onFailure: Callback<DataAccessException>) =
            this.participants.checkInParticipant(participant, onSuccess, onFailure)

    /**
     * @see ParticipantService
     */
    @Throws(DataAccessException::class)
    fun undoCheckInParticipant(participant: Participant): Participant =
            this.participants.undoCheckInParticipant(participant)

    /**
     * @see ParticipantService
     */
    fun undoCheckInParticipant(participant: Participant, onSuccess: Callback<Participant>, onFailure: Callback<DataAccessException>) =
            this.participants.undoCheckInParticipant(participant, onSuccess, onFailure)

    /**
     * @see ParticipantService
     */
    @Throws(DataAccessException::class)
    fun deleteParticipant(participant: Participant): Participant =
            this.participants.deleteParticipant(participant)

    /**
     * @see ParticipantService
     */
    fun deleteParticipant(participant: Participant, onSuccess: Callback<Participant>, onFailure: Callback<DataAccessException>) =
            this.participants.deleteParticipant(participant, onSuccess, onFailure)

    /**
     * @see ParticipantService
     */
    @Throws(DataAccessException::class)
    fun randomizeParticipants(tournament: Tournament): List<Participant> =
            this.participants.randomizeParticipants(tournament)

    /**
     * @see ParticipantService
     */
    fun randomizeParticipants(tournament: Tournament, onSuccess: Callback<List<Participant>>, onFailure: Callback<DataAccessException>) =
            this.participants.randomizeParticipants(tournament, onSuccess, onFailure)

    /**
     * @see MatchService
     */
    @JvmOverloads
    @Throws(DataAccessException::class)
    fun getMatches(tournament: Tournament, participant: Participant? = null, state: MatchState? = null): List<Match> =
            this.matches.getMatches(tournament, participant, state)

    /**
     * @see MatchService
     */
    @JvmOverloads
    fun getMatches(tournament: Tournament, participant: Participant? = null, state: MatchState? = null,
                   onSuccess: Callback<List<Match>>, onFailure: Callback<DataAccessException>) =
            this.matches.getMatches(tournament, participant, state, onSuccess, onFailure)

    /**
     * @see MatchService
     */
    @JvmOverloads
    @Throws(DataAccessException::class)
    fun getMatch(tournament: Tournament, matchId: Long, includeAttachments: Boolean = false): Match =
            this.matches.getMatch(tournament, matchId, includeAttachments)

    /**
     * @see MatchService
     */
    @JvmOverloads
    fun getMatch(tournament: Tournament, matchId: Long, includeAttachments: Boolean = false,
                 onSuccess: Callback<Match>, onFailure: Callback<DataAccessException>) =
            this.matches.getMatch(tournament, matchId, includeAttachments, onSuccess, onFailure)

    /**
     * @see MatchService
     */
    @Throws(DataAccessException::class)
    fun updateMatch(match: Match, data: MatchQuery): Match = this.matches.updateMatch(match, data)

    /**
     * @see MatchService
     */
    fun updateMatch(match: Match, data: MatchQuery, onSuccess: Callback<Match>, onFailure: Callback<DataAccessException>) =
            this.matches.updateMatch(match, data, onSuccess, onFailure)

    /**
     * @see MatchService
     */
    @Throws(DataAccessException::class)
    fun reopenMatch(match: Match): Match = this.matches.reopenMatch(match)

    /**
     * @see MatchService
     */
    fun reopenMatch(match: Match, onSuccess: Callback<Match>, onFailure: Callback<DataAccessException>) =
            this.matches.reopenMatch(match, onSuccess, onFailure)

    /**
     * @see AttachmentService
     */
    @Throws(DataAccessException::class)
    fun getAttachments(match: Match): List<Attachment> = this.attachments.getAttachments(match)

    /**
     * @see AttachmentService
     */
    fun getAttachments(match: Match, onSuccess: Callback<List<Attachment>>, onFailure: Callback<DataAccessException>) =
            this.attachments.getAttachments(match, onSuccess, onFailure)

    /**
     * @see AttachmentService
     */
    @Throws(DataAccessException::class)
    fun getAttachment(match: Match, attachmentId: Long): Attachment =
            this.attachments.getAttachment(match, attachmentId)

    /**
     * @see AttachmentService
     */
    fun getAttachment(match: Match, attachmentId: Long, onSuccess: Callback<Attachment>, onFailure: Callback<DataAccessException>) =
            this.attachments.getAttachment(match, attachmentId, onSuccess, onFailure)

    /**
     * @see AttachmentService
     */
    @Throws(DataAccessException::class)
    fun createAttachment(match: Match, data: AttachmentQuery): Attachment =
            this.attachments.createAttachment(match, data)

    /**
     * @see AttachmentService
     */
    fun createAttachment(match: Match, data: AttachmentQuery, onSuccess: Callback<Attachment>,
                         onFailure: Callback<DataAccessException>) =
            this.attachments.createAttachment(match, data, onSuccess, onFailure)

    /**
     * @see AttachmentService
     */
    @Throws(DataAccessException::class)
    fun updateAttachment(match: Match, attachment: Attachment, data: AttachmentQuery): Attachment =
            this.attachments.updateAttachment(match, attachment, data)

    /**
     * @see AttachmentService
     */
    fun updateAttachment(match: Match, attachment: Attachment, data: AttachmentQuery, onSuccess: Callback<Attachment>,
                         onFailure: Callback<DataAccessException>) =
            this.attachments.updateAttachment(match, attachment, data, onSuccess, onFailure)

    /**
     * @see AttachmentService
     */
    @Throws(DataAccessException::class)
    fun deleteAttachment(match: Match, attachment: Attachment): Attachment =
            this.attachments.deleteAttachment(match, attachment)

    /**
     * @see AttachmentService
     */
    fun deleteAttachment(match: Match, attachment: Attachment, onSuccess: Callback<Attachment>,
                         onFailure: Callback<DataAccessException>) =
            this.attachments.deleteAttachment(match, attachment, onSuccess, onFailure)
}